### PR TITLE
Fill in coverage gaps for Component2

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
### Motivation
There were missing tests for some of the new Component2 code.

### Changes
- Add full test coverage for `JsBackedMap` methods
    - Fix hashCode in DDC
- Add missing test coverage for `setStateWithUpdater` arguments
- Add tests for Component2Bridge
- Add tests for throwing `UnsupportedError`, fix missing `throws` keywords

Boyscouting:
- Make `bridgeForComponent` protected.

### Testing Instrructions
- Ensure CI passes

### CODE REVIEW

File: test/lifecycle_test.dart [hunk 1]
```diff
@@ -906,44 +906,84 @@ void sharedLifecycleTests<T extends react.Component>({
...
+        expectedSnapshot = null;
 
-      final Map expectedProps = unmodifiableMap(defaultProps, initialProps, emptyChildrenProps);
+        component = getDartComponent(render(LifecycleTest(initialProps)));
+        component.lifecycleCalls.clear();
+      });
 
-      final Null expectedSnapshot = null;
+      tearDown(() {
+        expect(
```
Comment (greglittlefield-wf) [inline test/lifecycle_test.dart]: This expectation is the same as before, just indented and placed in a `tearDown`.

File: test/lifecycle_test.dart [hunk 2]
```diff
@@ -906,44 +906,84 @@ void sharedLifecycleTests<T extends react.Component>({
...
+      dynamic newContext;
+      Null expectedSnapshot;
+      LifecycleTestHelper component;
 
-      final dynamic newContext = isComponent2 ? null : const {};
+      setUp(() {
+        initialProps = unmodifiableMap({'getInitialState': (_) => initialState});
+        expectedProps = unmodifiableMap(defaultProps, initialProps, emptyChildrenProps);
+        newContext = isComponent2 ? null : const {};
+        expectedSnapshot = null;
```
Comment (greglittlefield-wf) [inline test/lifecycle_test.dart]: These variables are the same as before; their initialization is just split to be able to put them in a `setUp`

File: test/react_client/bridge_test.dart [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  // Test behavior that isn't functionally tested via component lifecycle tests.
+  group('Component2Bridge', () {
+    group('.forComponent returns the bridge used by that mounted component, which', () {
+      test('by default is a const instance of Component2BridgeImpl', () {
+        final instance = getDartComponent<_Foo>(render(Foo({})));
+        expect(Component2Bridge.forComponent(instance), same(const Component2BridgeImpl()));
+      });
+
+      test('can be a custom bridge specified via registerComponent2', () {
+        customBridgeCalls.clear();
```
Comment (aaronlademann-wf) [inline test/react_client/bridge_test.dart]: Should this be put into a `setUp`?

Comment (greglittlefield-wf) [reply test/react_client/bridge_test.dart]: Since it's only used in this one test, I didn't think it was necessary.

Comment (greglittlefield-wf) [reply test/react_client/bridge_test.dart]: I'll improve this, though

Comment (greglittlefield-wf) [reply test/react_client/bridge_test.dart]: Added https://github.com/cleandart/react-dart/pull/208/commits/57bfbfd90d70e6f302fbc91a7c1c238d50d1c156 which I think is cleaner than what I had before

**Comment by aaronlademann-wf** [comment]
The failures are in the dev phase of the build (using Dart 2.6.0-dev)... which has a bug causing the failures: https://github.com/dart-lang/sdk/issues/38311

+10 to integration _(CI Passing on Dart 2 stable)_

**Pull request merged**